phpunit/class-rest-annotations-controller-test.php

```diff
@@ -0,0 +1,1244 @@
+<?php
```

Comment:
I haven't gone through all the tests here, but I am wondering if you have created test cases to handle all the parameter values and types that are passed to the API. For example, do you check the expected return values when a required parameter is missing, when a wrong value is passed in the request, whether errors are communicated correctly, etc? For example, I get back a status `200 OK` with a value of `Unexpected '<'` in the body of the response when I pass the parameters:

```
{
    "post": 1,
    "via": "gutenberg",
    "content": " ",
    "author": "123"
}
```

Is this normal, am I doing something wrong? Obviously `author: "123"` erred there.

Similarly, passing an empty string for content `content: ""` results in a `400 Bad Request`, while if I pass just a space with `content: " "` results in `200 OK`. I can see though in the response that the `rendered` content is null/empty `""`. Why allow that then and not treat it like the former case? etc..

Reply:
@chriskmnds Awesome work, thank you! 👁  
It's great to have another pair of eyes on this.

> test cases to handle all the parameter values and types that are passed to the API.

Yes, there were *some* like this already, but most had a focus on user permissions. I've been in the process of restructuring the unit tests for this branch and that work is done now. I just pushed the improved unit tests to this PR and I'll post a summary shortly.

---

Both great finds, and both bugs. Nice work!

- [x] 🐛  **"Unexpected '<' in the body of the response"**... Triggered by an invalid data type in the `author` parameter. This was resulting in an internal `WP_Error` response that went unhandled, resulting in a parse error and a malformed response instead of the error message and code. Fixed by adding [these three lines](https://github.com/jaswrks/gutenberg/blob/f95b0661f9785aa609317502a19527f75027644b/lib/class-wp-rest-annotations-controller.php#L373-L375) and unit tests [now cover this](https://github.com/WordPress/gutenberg/pull/4685/commits/af3b4c1b9de87b3b2edc38b2a0cd28134c0d9333).

- [x] 🐛  **"if I pass just a space with content: " " results in 200 OK."** ... Yeah, I think it should be smarter than this. The same behavior seems to exist in WordPress core. I corrected this for annotations by adding a [smarter analysis](https://github.com/WordPress/gutenberg/pull/4685/commits/f95b0661f9785aa609317502a19527f75027644b#diff-df5e27f1866a42a0c24ae7c0b17b1bc6R1291) of the content and [unit tests](https://github.com/jaswrks/gutenberg/blob/af3b4c1b9de87b3b2edc38b2a0cd28134c0d9333/phpunit/annotations/class-rest-annotations-controller-test.php#L884-L889) also.

Reply:
Thanks again Chris. I really appreciate you taking the time to look over this with me. Anything I can help you with just hit me on Slack! :-)